src/elements/content-sidebar/ActivitySidebar.js

```diff
@@ -827,6 +875,149 @@ class ActivitySidebar extends React.PureComponent<Props, State> {
+        if (!itemId) {
+            return undefined;
+        }
+        return feedItems.find(({ id }) => id === itemId);
+    };
+
+    /**
+     * Returns parent feed item based on the reply id
+     *
+     * @param {FeedItems} feedItems - the feed items
+     * @param {string} replyId - feed item's reply id
+     * @return {FeedItem | undefined}
+     */
+    getFeedItemByReplyId = (feedItems: FeedItems, replyId?: string): FeedItem | typeof undefined => {
+        if (!replyId) {
+            return undefined;
+        }
+        return feedItems.find(item => {
+            if (
+                (item.type !== FEED_ITEM_TYPE_ANNOTATION && item.type !== FEED_ITEM_TYPE_COMMENT) ||
+                !item.replies ||
+                !Array.isArray(item.replies)
+            ) {
+                return false;
+            }
+            return item.replies.some(({ id }) => id === replyId);
+        });
+    };
+
+    /**
+     * Returns true if item (based on given item id) is found within feed items or its replies
+     *
+     * @param {FeedItems} feedItems - the feed items
+     * @param {string} itemId - feed item id
+     * @return {boolean}
+     */
+    isItemInFeed = (feedItems: FeedItems, itemId: string): boolean =>
+        !!(this.getFeedItemById(feedItems, itemId) || this.getFeedItemByReplyId(feedItems, itemId));
+
+    /**
+     * Returns active entry data (id, type) based on the activeFeedEntryId and activeFeedEntryType values
+     * (it can be existing item or parent if the active entry id belongs to a reply)
+     *
+     * @param {FeedItems} feedItems - the feed items
+     * @return {Promise<{ id: string, type?: FocusableFeedItemType }>}
+     */
+    getActiveFeedEntryData = (feedItems: FeedItems): Promise<{ id?: string, type?: FocusableFeedItemType }> => {
+        const { activeFeedEntryId, activeFeedEntryType } = this.props;
+        return new Promise((resolve, reject) => {
+            if (
+                !activeFeedEntryId ||
+                (activeFeedEntryType !== FEED_ITEM_TYPE_ANNOTATION &&
+                    activeFeedEntryType !== FEED_ITEM_TYPE_COMMENT &&
+                    activeFeedEntryType !== FEED_ITEM_TYPE_TASK)
+            ) {
+                resolve({ id: activeFeedEntryId, type: activeFeedEntryType });
+                return;
+            }
+
+            // Check if the active entry is a first level Feed item
+            const firstLevelItem = this.getFeedItemById(feedItems, activeFeedEntryId);
+            if (firstLevelItem) {
+                const { id, type } = firstLevelItem;
+                if (
+                    type === FEED_ITEM_TYPE_ANNOTATION ||
+                    type === FEED_ITEM_TYPE_COMMENT ||
+                    type === FEED_ITEM_TYPE_TASK
+                ) {
+                    resolve({ id, type });
+                    return;
+                }
+                resolve({ id: activeFeedEntryId, type: activeFeedEntryType });
+                return;
```

Comment:
We should avoid multiple returns, as it makes logic very difficult to debug and reason about. We're doing quite a bit of duplicate work in this function and it would be great to refactor it for simplicity.